pkg/csi/service/vanilla/controller.go

```diff
@@ -906,6 +921,464 @@ func (c *controller) getAccessibleTopologiesForDatastore(ctx context.Context, vc
+			Names: []string{string(cnstypes.QuerySelectionNameTypeDataStoreUrl)},
+		}
+		queryResult, err := utils.QueryVolumeUtil(ctx, volumeMgr, queryFilter, &querySelection, true)
+		if err != nil {
+			// TODO: QueryVolume need to return faultType.
+			// Need to return faultType which is returned from QueryVolume.
+			// Currently, just return "csi.fault.Internal".
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"queryVolumeUtil failed for volumeID: %s in vCenter %q. Error: %+v",
+				volumeInfo.VolumeID.Id, vcHost, err)
+		}
+		if len(queryResult.Volumes) == 0 || queryResult.Volumes[0].DatastoreUrl == "" {
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"queryVolumeUtil could not retrieve volume information for volume ID: %q in vCenter %q",
+				volumeInfo.VolumeID.Id, vcHost)
+		}
+		datastoreURL = queryResult.Volumes[0].DatastoreUrl
+	}
+
+	// Retrieve datastore topology information from CSINodeTopology CRs.
+	allNodeVMs, err = c.nodeMgr.GetAllNodesByVC(ctx, vcHost)
+	if err != nil {
+		return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to fetch VirtualMachines for the registered nodes in VC %q. Error: %v", vcHost, err)
+	}
+	// Find datastore topology from the retrieved datastoreURL.
+	datastoreAccessibleTopology, err = c.calculateAccessibleTopologiesForDatastore(ctx, vcenter,
+		vcTopologySegmentsMap[vcHost], allNodeVMs, datastoreURL)
+	if err != nil {
+		return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to calculate accessible topologies for the datastore %q", datastoreURL)
+	}
+
+	// Add topology segments to the CreateVolumeResponse.
+	for _, topoSegments := range datastoreAccessibleTopology {
+		volumeTopology := &csi.Topology{
+			Segments: topoSegments,
+		}
+		resp.Volume.AccessibleTopology = append(resp.Volume.AccessibleTopology, volumeTopology)
+	}
+
+	// Set the Snapshot VolumeContentSource in the CreateVolumeResponse
+	if contentSourceSnapshotID != "" {
+		resp.Volume.ContentSource = &csi.VolumeContentSource{
+			Type: &csi.VolumeContentSource_Snapshot{
+				Snapshot: &csi.VolumeContentSource_SnapshotSource{
+					SnapshotId: contentSourceSnapshotID,
+				},
+			},
+		}
+	}
+	// Create CNSVolumeInfo CR for the volume ID.
+	err = volumeInfoService.CreateVolumeInfo(ctx, volumeInfo.VolumeID.Id, vcHost)
+	if err != nil {
+		return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to store volumeID %q for vCenter %q in CNSVolumeInfo CR. Error: %+v",
+			volumeInfo.VolumeID.Id, vcHost, err)
+	}
+	return resp, "", nil
+}
+
+// calculateAccessibleTopologiesForDatastore figures out the list of topologies from
+// which the given datastore is accessible when multi-VC FSS is enabled.
+func (c *controller) calculateAccessibleTopologiesForDatastore(ctx context.Context, vcenter *cnsvsphere.VirtualCenter,
```

Comment:
calculateAccessibleTopologiesForDatastore => getAccessibleTopologiesForDatastore

Reply:
There is already an older function named getAccessibleTopologiesForDatastore. I had to create a new one for Multi-VC, so named it calculate